### PR TITLE
Generate client factory for arm

### PR DESCRIPTION
**Pull request by tadelesh**
Resolve: https://github.com/Azure/autorest.go/issues/823.
For all 2364 mgmt. clients, there are only five modules with 42 clients that have additional client parameters besides subscription ID. After reviewed all of them, I believe most of them are from wrong swagger definition, and I will double check with service team and use directive to fix them in their next version.

### CODE REVIEW

**Comment by tadelesh** [comment]
There is one thing I did not achieve for now: deep copy options before client creation check. My idea is to add the function to `azcore/arm` package.

File: test/compute/2019-12-01/armcompute/zz_client_factory.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+// DO NOT EDIT.
+
+package armcompute
+
+import (
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+)
+
+// ClientFactory is a factory help to create all clients in this modules.
```
Comment (jhendrixMSFT) [inline test/compute/2019-12-01/armcompute/zz_client_factory.go:17]:
```go
// ClientFactory is a client factory used to create any client in this module.
// Don't use this type directly, use NewClientFactory instead.
```

File: test/compute/2019-12-01/armcompute/zz_client_factory.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+type ClientFactory struct {
+	subscriptionID string
+	credential     azcore.TokenCredential
+	options        *arm.ClientOptions
+}
+
+// NewClientFactory creates a new instance of ClientFactor with the specified values.
```
Comment (jhendrixMSFT) [inline test/compute/2019-12-01/armcompute/zz_client_factory.go:24]:
```go
// NewClientFactory creates a new instance of ClientFactory with the specified values.
// The parameter values will be propagated to any client created from this factory.
```

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
Some minor refinements on the doc comments.

**Comment by jhendrixMSFT** [comment]
@JeffreyRichter here is the client factory codegen.

**Comment by JeffreyRichter** [comment]
> There is one thing I did not achieve for now: deep copy options before client creation check because it is not guaranteed that nothing change for options after check. My idea is to add a deep copy function for options to `azcore/arm` package.

I don't think we should need this. Just make a copy with *options because it shouldn't contain any pointers. Plus, client ctors should all be making copies to ensure that any customer code modifications to an options structure does NOT impact any clients already constructed.

**Comment by tadelesh** [comment]
> > There is one thing I did not achieve for now: deep copy options before client creation check because it is not guaranteed that nothing change for options after check. My idea is to add a deep copy function for options to `azcore/arm` package.
> 
> I don't think we should need this. Just make a copy with *options because it shouldn't contain any pointers. Plus, client ctors should all be making copies to ensure that any customer code modifications to an options structure does NOT impact any clients already constructed.

There are some slices inside the options (e.g., `PerCallPolicies`, `PerRertryPolicies`). Though changing the element in a slice should be a rare usage, it could be possible that options will be changed after factory created.

File: test/databoxedge/2021-02-01/armdataboxedge/zz_client_factory.go [hunk 3]
```diff
@@ -32,106 +34,106 @@ func NewClientFactory(subscriptionID string, credential azcore.TokenCredential,
 	}
 	return &ClientFactory{
 		subscriptionID: subscriptionID, credential: credential,
-		options: options,
+		options: *options,
```
Comment (jhendrixMSFT) [inline test/databoxedge/2021-02-01/armdataboxedge/zz_client_factory.go:37]: This will panic if `options` is `nil`.

**Comment by JeffreyRichter** [comment]
> > > There is one thing I did not achieve for now: deep copy options before client creation check because it is not guaranteed that nothing change for options after check. My idea is to add a deep copy function for options to `azcore/arm` package.
> > 
> > 
> > I don't think we should need this. Just make a copy with *options because it shouldn't contain any pointers. Plus, client ctors should all be making copies to ensure that any customer code modifications to an options structure does NOT impact any clients already constructed.
> 
> There are some slices and map inside the options (e.g., `PerCallPolicies`, `PerRertryPolicies`, `Services` in `Cloud.Configuration`). Though changing the element in a slice or delete a map element should be a rare usage for options, it could be possible after factory created. And the change of cloud configuration bypass the check in factory CTOR. Shall we handle it?

Yes, you're right. make a deep copy.

File: test/databoxedge/2021-02-01/armdataboxedge/zz_client_factory.go (hunk 3 above)
Comment (jhendrixMSFT) [reply test/databoxedge/2021-02-01/armdataboxedge/zz_client_factory.go:37]: You can add the `nil` handling in the `Clone()` impl.

**Review by xboxeer — APPROVED** [review]
Had an offline discussion with Chenjie, the new API LGTM

File: test/compute/2019-12-01/armcompute/zz_client_factory.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+//   - credential - used to authorize requests. Usually a credential from azidentity.
+//   - options - pass nil to accept the default values.
+func NewClientFactory(subscriptionID string, credential azcore.TokenCredential, options *arm.ClientOptions) (*ClientFactory, error) {
+	_, err := arm.NewClient("armcompute.ClientFactory", moduleVersion, credential, options)
+	if err != nil {
+		return nil, err
+	}
+	return &ClientFactory{
+		subscriptionID: subscriptionID, credential: credential,
+		options: options.Clone(),
```
Comment (tadelesh) [inline test/compute/2019-12-01/armcompute/zz_client_factory.go:38]: Add `Clone` logic. @jhendrixMSFT please help to review again. Thanks.

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**